File: src/portfolio_manager/services/exchange_rate_service.py
"""Service for FX rates used in portfolio valuation."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, timedelta
from decimal import Decimal

from portfolio_manager.services.exim_exchange_rate_client import (
    EximExchangeRateClient,
)


@dataclass(frozen=True)
class ExchangeRateService:
    """Provide USD/KRW exchange rate from a fixed value or EXIM client."""

    exim_client: EximExchangeRateClient | None = None
    fixed_usd_krw_rate: Decimal | None = None

    def get_usd_krw_rate(self, search_date: str | None = None) -> Decimal:
        """Return USD/KRW rate as Decimal."""
        if self.fixed_usd_krw_rate is not None:
            return self.fixed_usd_krw_rate
        if self.exim_client is None:
            raise ValueError("Exchange rate source is not configured")
        if search_date is None:
            base_date = date.today()
            for offset in range(0, 7):
                candidate = (base_date - timedelta(days=offset)).strftime("%Y%m%d")
                try:
                    rate = self.exim_client.fetch_usd_rate(search_date=candidate)
                except ValueError as exc:
                    if str(exc) != "USD rate not found":
                        raise
                    continue
                return Decimal(str(rate))
            raise ValueError("USD rate not found")
        rate = self.exim_client.fetch_usd_rate(search_date=search_date)
        return Decimal(str(rate))

File: internal/services/exchange_rate_service.go
package services

import (
	"fmt"
	"sync"

	"github.com/kadragon/portfolio-manager/internal/datex"
	"github.com/kadragon/portfolio-manager/internal/ktime"
	"github.com/kadragon/portfolio-manager/internal/numeric"
)

// EximRateClient fetches the USD/KRW rate from the Korea EXIM Bank API.
type EximRateClient interface {
	FetchUSDRate(searchDate string) (float64, error)
}

// ExchangeRateService provides USD/KRW exchange rates.
// Supports a fixed rate or a live EXIM client with 7-day backoff and in-memory cache.
type ExchangeRateService struct {
	mu          sync.Mutex
	fixedRate   *numeric.Decimal
	eximClient  EximRateClient
	cachedRates map[string]numeric.Decimal
}

// NewFixedExchangeRateService creates a service with a fixed USD/KRW rate.
func NewFixedExchangeRateService(usdKRW numeric.Decimal) *ExchangeRateService {
	return &ExchangeRateService{fixedRate: &usdKRW}
}

// NewEximExchangeRateService creates a service that fetches live USD/KRW rates from EXIM.
func NewEximExchangeRateService(client EximRateClient) *ExchangeRateService {
	return &ExchangeRateService{
		eximClient:  client,
		cachedRates: make(map[string]numeric.Decimal),
	}
}

// GetUSDKRW returns the USD/KRW rate.
// Fixed-rate services return immediately; EXIM services try today then up to 6 prior days.
func (s *ExchangeRateService) GetUSDKRW() numeric.Decimal {
	if s.fixedRate != nil {
		return *s.fixedRate
	}
	if s.eximClient == nil {
		return numeric.Zero
	}

	today := datex.FromTime(ktime.NowKST())
	for offset := 0; offset < 7; offset++ {
		candidate := today.Time.AddDate(0, 0, -offset).Format("20060102")
		s.mu.Lock()
		if v, ok := s.cachedRates[candidate]; ok {
			s.mu.Unlock()
			return v
		}
		s.mu.Unlock()

		raw, err := s.eximClient.FetchUSDRate(candidate)
		if err != nil || raw == 0 {
			continue
		}
		rate, parseErr := numeric.FromString(fmt.Sprintf("%g", raw))
		if parseErr != nil {
			continue
		}
		s.mu.Lock()
		s.cachedRates[candidate] = rate
		s.mu.Unlock()
		return rate
	}
	return numeric.Zero
}
